### PR TITLE
2.0.0 beta2

### PR DESCRIPTION
**Pull request by MatMaul**
This should keep backward compatibility with both distributor and connector I believe.

EDIT: well actually it seems to also be the case on option 3 that @p1gp1g implemented, I went a bit fast ^^. So the proposal are pretty similar and mostly differ in naming, using an extra instead of an action connector side and trying to avoid splitting the spec in v1/v2 documents.

### CODE REVIEW

File: specifications/android.md [hunk 1]
```diff
@@ -21,8 +21,10 @@ The distributor MUST expose a broadcast receiver with the following actions:
 * org.unifiedpush.android.distributor.REGISTER
 * org.unifiedpush.android.distributor.UNREGISTER
 
-This broadcast receiver is the Registration Broadcast Receiver.
+The distributor MUST also expose the following action if it supports sending bytes messages:
+* org.unifiedpush.android.distributor.feature.bytesMessage
```
Comment (binwiederhier) [inline specifications/android.md:25]: feature.BYTES_MESSAGE

File: specifications/android.md [hunk 2]
```diff
@@ -49,18 +51,21 @@ The exposed broadcast receiver of the distributor application MUST handle 2 diff
...
 
-The connector send this action to register to push messages. The intent MUST contain 2 String extras:
-* application: with the end user application package name. The distributor MUST be able to handle many connections with a single application.
-* token: with a random token to identify the connection between the connector and the distributor. It MUST be unique on distributor side.
+The connector send this action to register to push messages. The intent MUST contain 2 extras:
+* application (String): with the end user application package name. The distributor MUST be able to handle many connections with a single application.
+* token (String): with a random token to identify the connection between the connector and the distributor. It MUST be unique on distributor side.
+
+It MAY be send with the following extra:
+* features (ArrayList<String>): indicate the connector supports a set of optional features. Only "bytesMessage" is currently specified.
```
Comment (binwiederhier) [inline specifications/android.md:59]: Should this be the FQCN of the feature?

File: specifications/android.md [hunk 3]
```diff
@@ -49,18 +51,21 @@ The exposed broadcast receiver of the distributor application MUST handle 2 diff
 
 ### org.unifiedpush.android.distributor.REGISTER
 
-The connector send this action to register to push messages. The intent MUST contain 2 String extras:
-* application: with the end user application package name. The distributor MUST be able to handle many connections with a single application.
-* token: with a random token to identify the connection between the connector and the distributor. It MUST be unique on distributor side.
+The connector send this action to register to push messages. The intent MUST contain 2 extras:
```
Comment (binwiederhier) [inline specifications/android.md:54]: The connector sends

File: specifications/android.md [hunk 4]
```diff
@@ -82,32 +87,38 @@ There is a third action the connector SHOULD handle:
...
 The distributor MUST send this action to the registered application to forward a push message received from the provider to the end user application.
 
-It MUST be send with the 2 following extras:
-* token: the token supplied by the end user application during registration
-* message: the push message sent by the application server. It MUST be the raw POST data received by the rewrite proxy.
+It MUST be send with the following extra:
+* token (String): the token supplied by the end user application during registration
+
+It MUST be send with at least one of the 2 following extras:
+* message (String): the push message sent by the application server, as a string. It MUST be setted if the connector has been registered without support for bytesMessage feature. If setted, it MUST be the raw POST data received by the rewrite proxy.
```
Comment (binwiederhier) [inline specifications/android.md:109]: It is ambiguous if the distributor has to set this if connected with the feature.

I would say: It MUST be set if the bytesMessage feature was not present, and it MAY be present if it was present during registration.

Grammar: "If set, ..." Not "If setted"

File: specifications/android.md (hunk 1 above)
Comment (p1gp1g) [reply specifications/android.md:25]: So connector can check if the feature it wants is supported with this instead of v2. Nice.

File: specifications/android.md (hunk 2 above)
Comment (p1gp1g) [reply specifications/android.md:59]: The supported features should be listed somewhere else I think

File: specifications/android.md [hunk 5]
```diff
@@ -82,32 +87,38 @@ There is a third action the connector SHOULD handle:
...
 
 The distributor MUST send this action to the registered application to confirm the registration of an end user application, when a registered application send again an action with the intent org.unifiedpush.android.distributor.REGISTER and a valid token, or when the endpoint change with the 2 following extras:
-* token: the token supplied by the end user application during registration
-* endpoint: the endpoint
+* token (String): the token supplied by the end user application during registration
+* endpoint (String): the endpoint
 
 ### org.unifiedpush.android.connector.REGISTRATION_FAILED
 
 The distributor MUST send this action to the registered application if the token is already registered for another application or if the registration can not be processed (for instance when the distributor is not connected to its provider server) with the 2 following extras:
```
Comment (p1gp1g) [inline specifications/android.md:95]: If the feature is not implemented too

**Review by p1gp1g — COMMENTED** [review]
To be future-proof, distributors must know how to handle features they don't know : if a distributor receives a feature it doesn't know, it must send a registration failed.

Also we may have optional feature vs required feature : the extra can be split in two extras : required_features and optional_features then the distributor must ack somehow the optional features - which can be done on new endpoint. On my side, to keep things simple I'd argue for required-only :)

--- 

Very nice proposal. The features extra can be very extensible and the connector can easily know if the distributor has the feature before trying to register. Let's go for it. It just requires a few rewording.

PS: we need to think about platform-specific features for flutter-connector now

**Comment by MatMaul** [comment]
> To be future-proof, distributors must know how to handle features they don't know : if a distributor receives a feature it doesn't know, it must send a registration failed.

It was to keep full backward compatibility for distributors too, in my proposal message needs to be handled as a fallback by the client. But after thinking it's indeed a good idea that the connector knows if it supports a feature at registration time, because if it's required for the app to work it's sad to get an error only on first message...

> .... On my side, to keep things simple I'd argue for required-only :)

Agreed :+1:

File: specifications/android.md [hunk 6]
```diff
@@ -82,32 +91,43 @@ There is a third action the connector SHOULD handle:
...
-* token: the token supplied by the end user application during registration
-* message: the push message sent by the application server. It MUST be the raw POST data received by the rewrite proxy.
+It MUST be send with the following extra:
+* token (String): the token supplied by the end user application during registration
+
+It MUST be send with at least one of the 2 following extras:
+* message (String): the push message sent by the application server, as a string. It MUST be set if the BYTES_MESSAGE feature was not requested during registration. If set, it MUST be the raw POST data received by the rewrite proxy.
+* bytesMessage (ByteArray): the push message sent by the application server, as an array of bytes. It MAY be ommited. If set, it MUST be the raw POST data received by the rewrite proxy.
+
+The distributor MAY set message on top of bytesMessage if BYTES_MESSAGE feature was requested.
```
Comment (binwiederhier) [inline specifications/android.md:122]: While correct, the paragraph is hard to parse. May I suggest something like this:

```
If the BYTES_MESSAGE feature was requested, it MUST send the following extras:
* token (String): the token supplied by the end user application during registration
* bytesMessage (ByteArray): the push message sent by the application server, as an array of bytes.

If the BYTES_MESSAGE feature was requested, it MAY additionally send the message as a string:
* message (String): the push message sent by the application server, as a string.

If the BYTES_MESSAGE feature was not requested, it MUST send the following extras:
* token (String): the token supplied by the end user application during registration
* message (String): the push message sent by the application server, as a string.
```

File: specifications/android.md [hunk 7]
```diff
@@ -49,18 +55,21 @@ The exposed broadcast receiver of the distributor application MUST handle 2 diff
...
 ### org.unifiedpush.android.distributor.REGISTER
 
-The connector send this action to register to push messages. The intent MUST contain 2 String extras:
-* application: with the end user application package name. The distributor MUST be able to handle many connections with a single application.
-* token: with a random token to identify the connection between the connector and the distributor. It MUST be unique on distributor side.
+The connector sends this action to register to push messages. The intent MUST contain 2 extras:
+* application (String): with the end user application package name. The distributor MUST be able to handle many connections with a single application.
+* token (String): with a random token to identify the connection between the connector and the distributor. It MUST be unique on distributor side.
+
+It MAY be send with the following extra:
```
[the remark this answers is not in the document]
Comment (karmanyaahm) [reply specifications/android.md:62]: I think you meant 'sent'?

File: specifications/android.md (hunk 6 above)
Comment (p1gp1g) [reply specifications/android.md:122]: We should add the message (String) is not the full raw POST data. It works only if the message contains characters with char code between 0 and 127

Comment (MatMaul) [reply specifications/android.md:124]: I've kept the req about RAW post data for the bytesMessage use case, and put nothing for the string version.

**Review by p1gp1g — APPROVED** [review]
OK for me !

File: specifications/android.md [hunk 8]
```diff
@@ -82,32 +91,45 @@ There is a third action the connector SHOULD handle:
...
 
-It MUST be send with the 2 following extras:
-* token: the token supplied by the end user application during registration
-* message: the push message sent by the application server. It MUST be the raw POST data received by the rewrite proxy.
+If the BYTES_MESSAGE feature was requested, it MUST send the following extras:
+* token (String): the token supplied by the end user application during registration
+* bytesMessage (ByteArray): the push message sent by the application server, as an array of bytes. It MUST be the raw POST data received by the rewrite proxy.
+
+If the BYTES_MESSAGE feature was requested, it MAY additionally send the message as a string:
+* message (String): the push message sent by the application server, as a string. This extra can't have characters with char code > 127.
```
Comment (binwiederhier) [inline specifications/android.md:119]: The char code >127 thing is really odd. Are you sure about that? That would mean that not even UTF-8 is supported, which is not true. 

Maybe this should just say that the message must be valid utf-8?

File: specifications/android.md [hunk 9]
```diff
@@ -49,18 +55,21 @@ The exposed broadcast receiver of the distributor application MUST handle 2 diff
...
 ### org.unifiedpush.android.distributor.REGISTER
 
-The connector send this action to register to push messages. The intent MUST contain 2 String extras:
-* application: with the end user application package name. The distributor MUST be able to handle many connections with a single application.
-* token: with a random token to identify the connection between the connector and the distributor. It MUST be unique on distributor side.
+The connector sends this action to register to push messages. The intent MUST contain 2 extras:
+* application (String): with the end user application package name. The distributor MUST be able to handle many connections with a single application.
+* token (String): with a random token to identify the connection between the connector and the distributor. It MUST be unique on distributor side.
+
+It MAY be sent with following extra:
```
Comment (sparchatus) [inline specifications/android.md:62]: Can we add another optional value?
* purpose (String): A short description of the purpose this registration fulfills which can be shown to the user by a distributor.

**Comment by sparchatus** [comment]
The [Android documentation](https://developer.android.com/guide/components/intents-filters) states
> If you need to declare your own extra keys (for intents that your app receives), be sure to include your app's package name as a prefix, as shown in the following example:
```kotlin
const val EXTRA_GIGAWATTS = "com.example.EXTRA_GIGAWATTS"
```
Should we try to change this or do we jut leave it since it works this way just as well?

File: specifications/android.md [hunk 10]
```diff
@@ -49,18 +55,21 @@ The exposed broadcast receiver of the distributor application MUST handle 2 diff
 
 ### org.unifiedpush.android.distributor.REGISTER
 
-The connector send this action to register to push messages. The intent MUST contain 2 String extras:
-* application: with the end user application package name. The distributor MUST be able to handle many connections with a single application.
-* token: with a random token to identify the connection between the connector and the distributor. It MUST be unique on distributor side.
+The connector sends this action to register to push messages. The intent MUST contain 2 extras:
+* application (String): with the end user application package name. The distributor MUST be able to handle many connections with a single application.
```
[the remark this answers is not in the document]
Comment (p1gp1g) [reply specifications/android.md:59]: The opposite sounds weird to me. Maybe because it is not my mother tongue

File: specifications/android.md (hunk 8 above)
Comment (p1gp1g) [reply specifications/android.md:119]: We should replace that with a point at the beginning saying all the String extras are UTF-8 encoded

File: specifications/android.md (hunk 9 above)
Comment (p1gp1g) [reply specifications/android.md:62]: I'm ok. I would just change "purpose" to "message" since it is quite the same meaning as "message" for registration failed.

* message (String) : A short description of the purpose of the registration that the connector MAY show to the user.

??

File: specifications/android.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (MatMaul) [reply specifications/android.md]: I removed the withs, it matches the rest of the spec. I also replaced `random` by `randomly generated` because `random` could be interpreted as `any` I believe.

**Comment by MatMaul** [comment]
> Should we try to change this or do we jut leave it since it works this way just as well?

This would break all backward compatibility so I would say we :see_no_evil:

File: specifications/android.md [hunk 11]
```diff
@@ -49,18 +61,22 @@ The exposed broadcast receiver of the distributor application MUST handle 2 diff
...
-* application: with the end user application package name. The distributor MUST be able to handle many connections with a single application.
-* token: with a random token to identify the connection between the connector and the distributor. It MUST be unique on distributor side.
+The connector sends this action to register to push messages. The intent MUST contain 2 extras:
+* application (String): the end user application package name. The distributor MUST be able to handle many connections with a single application.
+* token (String): a randomly generated token to identify the connection between the connector and the distributor. It MUST be unique on distributor side.
+
+It MAY be sent with the following 2 extras:
+* features (ArrayList<String>): indicate the connector is requesting a set of optional features to be enabled. It MUST be the qualified name of the action declared to advertise this feature. The connector MUST check that the action is declared before requesting an optional feature.
+* message (String) : a short description of the purpose of the registration that the connector MAY show to the user.
```
Comment (binwiederhier) [inline specifications/android.md:70]: 1. Remove space in front of the colon 
2. Should this be "that the distributor MAY  show to the user"?

File: specifications/android.md (hunk 8 above)
Comment (binwiederhier) [reply specifications/android.md:126]: My newly found knowledge about utf-8 tells me that only chars 127-192 are not allowed. I thought we said we'll remove this piece.

File: specifications/android.md (hunk 11 above)
Comment (MatMaul) [reply specifications/android.md]: indeed, thanks !

File: specifications/android.md (hunk 8 above)
Comment (MatMaul) [reply specifications/android.md]: Indeed I forgot to remove that after adding the general statement, done.

Comment (MatMaul) [reply specifications/android.md]: > My newly found knowledge about utf-8 tells me that only chars 127-192 are not allowed.

Should we put that as a restriction instead of UTF-8 ?

Comment (binwiederhier) [reply specifications/android.md]: I don't think so. It just confuses people. I also don't know if maybe be the encoding doesn't itself contain those characters. It is just disallowed in the first byte of a code point as far as I understand. Just remove it. I think the comment at the top covers everything.

Comment (MatMaul) [reply specifications/android.md]: I was talking about the comment at the top, this is already removed :) ok let's keep it like that.

**Review by sparchatus — APPROVED** [review]

**Review by p1gp1g — CHANGES_REQUESTED** [review]
We need to add all string extra must be utf-8 encoded; Then I think t is ready to merge

**Comment by MatMaul** [comment]
> We need to add all string extra must be utf-8 encoded; Then I think t is ready to merge

It's there, cf `General` section 🙂

**Review by p1gp1g — APPROVED** [review]
Sorry I missed that. Perfect !

**Review by binwiederhier — APPROVED** [review]

**Review by sparchatus — APPROVED** [review]
